TPM and Measured boot support SHA384 and SM3

lsubash opened the pull request:
Current Slimbootloader supports only TPM for SHA256 PCR
This patch introduces flexibility for user to select TPM
PCR bank available while build. Support of SHA384 and
SM3 to TPM measured boot are added.

For first boot SHA256/default TPM PCR would be available which
is default for most supported TPM's. Selected TPM PCR will be
available after reboot.

TPM selection would be based on config data and support would
be added in subsequent patches.

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

sagraw2 commented on BootloaderCommonPkg/BootloaderCommonPkg.dec:
```diff
@@ -222,6 +222,15 @@
 
+[PcdsDynamic]
+  ## This PCD indicates the hash algorithm to be enabled by bootloader for measured boot
+  #  Based on the value set, PCR bank world be enabled and extended
```
"This PCD indicates the PCR bank to be enabled/supported by Slim Bootloader for measured boot"

mauricema replied:
Do we really need dynamic PCD ?  
We are trying to void if possible since it will increase the binary size.
Can this be a fixed config ?

lsubash replied:
Platform configdata would define TPM PCR bank to be enabled. This dynamic pcd would be updated and used in TPMlib and necessary libs to check the PCR bank enabled. 
Incase if TPM Alg is NOT available in platform config data(for legacy platforms), default value in dynamic PDC would be used.

sagraw2 commented on PayloadPkg/OsLoader/OsLoader.c:
```diff
@@ -387,7 +387,7 @@ SetupBootImage (
       if (FeaturePcdGet (PcdMeasuredBootEnabled) && (LoaderPlatformInfo->LdrFeatures & FEATURE_MEASURED_BOOT)) {
         // Extend hash of the image into TPM.
-        TpmExtendPcrAndLogEvent (8, TPM_ALG_SHA256, LoadedImage->ImageHash,
+        TpmExtendPcrAndLogEvent (8, (TPMI_ALG_HASH) GetTpmHashAlgEnabled(), LoadedImage->ImageHash,
```
Will the ImageHash be of the same type as the PCR bank enabled in TPM?

lsubash replied:
It would be different in some cases. As OsLoader uses container support, I could add condition to check hash type.

sagraw2 commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -110,15 +116,33 @@ PreparePayload (
       if (ComponentEntry != NULL)  {
-        if (ComponentEntry->HashSize == SHA256_DIGEST_SIZE) {
-          TpmExtendHash (ComponentEntry->HashData, HASH_TYPE_SHA256);
+
```
Would it be cleaner to create a separate method in this file to move all the hash extend logic into it?

lsubash replied:
Yes, I would add separate method.

sagraw2 commented on BootloaderCommonPkg/Library/TpmLib/TpmLib.c:
```diff
@@ -17,11 +17,72 @@
+
+  PcdHashType = PcdGet8(PcdMeasuredBootHashType);
+
+  if(PcdHashType == HASH_TYPE_SHA256){
```
We have three different enums, HASH_TYPE_SHA256, HASH_ALG_SHA256 and TPM_ALG_SHA256.

I am afraid this will cause bugs later.
Can we atleast get rid of one?

lsubash replied:
Three hashes,
-  TCG Alg registry (bit mask) - Limited to TPM lib
-  TPM Alg registry - Limited to TPM Lib
- Another one is CryptoLib HASH_ALG_TYPE which is used across SBL.

We could think of aligning Crypto HASH_ALG_TYPE with TCG Alg registry. 
I would take this as separate one as it would involves many file changes including data type changes in API's.

mauricema commented on BootloaderCommonPkg/Library/TpmLib/TpmLib.c:
```diff
@@ -604,33 +682,51 @@ TpmIndicateReadyToBoot (
+        EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+    } else {
+      DEBUG ((DEBUG_INFO, "Calculate Hash for the components\n"));
+      Status = GetComponentInfo (Signature, &Src, &Length);
```
Can we consider doing it at the loading time ?   The normal flow will copy it to memory and then do hash calculation, it is more efficient if the component is in flash.

lsubash replied:
Thanks. I would check for doing hash calculation at load time and then extend.

lsubash replied:
Maurice, I will check this optimization as separate patch

mauricema commented:
For the reboot,  will it occur only once if the TPM bank is different ?

gdong1 commented on BootloaderCommonPkg/Library/TpmLib/TpmLib.c:
```diff
@@ -162,26 +223,34 @@ TpmPcrBankCheck (
+    // Check if TPM supports  PCR bank
+    if (!(TpmHashAlgorithmBitmap & HASH_ALG_SHA256) &&  !(TpmHashAlgorithmBitmap & HASH_ALG_SHA384)
+                                                    && !(TpmHashAlgorithmBitmap & HASH_ALG_SM3_256)){
+      DEBUG ((DEBUG_ERROR, "Hash algorithm is NOT supported. This should never happen with TPM2.0 device. \n"));
```
Better uses below to make it follow coding style and clear.
if ((TpmHashAlgorithmBitmap  & (HASH_ALG_SHA256 | HASH_ALG_SHA384 | HASH_ALG_SM3_256))  == 0) {

lsubash replied:
will update.

gdong1 commented on BootloaderCommonPkg/Library/TpmLib/TpmLib.c:
```diff
@@ -17,11 +17,72 @@
+**/
+UINT32
+GetTpmHashAlgEnabled (
+  VOID
```
Better let this function return TPMI_ALG_HASH or HASH_ALG_TYPE instead of UINT32?

sagraw2 replied:
Although SBL does NOT support multiple active PCR banks, there may be use-case when customer wants to support that. To keep our code scalable, let's keep this as UINT32 itself to reflect a bit mask.

lsubash replied:
This function would be updated to support MeasuredBoot Hashmask .

sagraw2 commented on BootloaderCommonPkg/Library/TpmLib/TpmEventLog.c:
```diff
@@ -259,11 +259,20 @@ TpmLogSpecIDEvent (
-
-    DigestSize = &DigestSize[1];
-    ++NumberOfAlgorithms;
+  } else if ((ActivePcr & HASH_ALG_SHA384) != 0) {
```
Let's replace 'else if' with 'if' here.
This will allow in future to support multiple PCR banks simultaneously.

Also, do NumberOfAlgorithms++ in every 'if' section.

lsubash replied:
yes sachin.

sagraw2 commented on BootloaderCommonPkg/Library/TpmLib/TpmEventLog.c:
```diff
@@ -376,6 +384,18 @@ TpmLogLocalityEvent (
     Digests->digests[Digests->count].hashAlg = TPM_ALG_SHA256;
     ZeroMem (& (Digests->digests[Digests->count].digest), GetHashSizeFromAlgo (TPM_ALG_SHA256));
     Digests->count = Digests->count + 1;
+  } else if ((ActivePcrBanks & HASH_ALG_SHA384) != 0) {
```
Same here. Replace 'else if' with 'if'.

lsubash commented on BootloaderCommonPkg/BootloaderCommonPkg.dec:
```diff
@@ -222,6 +222,15 @@
+  #     0x02    - SHA2_384.<BR>
+  #     0x03    - SHA2_512.<BR>
+  #     0x04    - SM3_256.<BR>
+  gPlatformCommonLibTokenSpaceGuid.PcdMeasuredBootHashType    |         0x1| UINT8  | 0x20000703
```
this will be updated to hashmask to support multiple PCR's

lsubash commented:
Updated patch with review comments. Also added following support,
- Enhance API's to support multiple PCR's. Currently only one Active PCR will be supported.
  Added following API's.
  GetCryptoHashAlg 
  TpmLibGetActivePcrBanks
  TpmHashAndExtendPcrEventLog
  TpmExtendOsHash
  
  Other API's were updated.
  
- Updated  PcdMeasuredBootHashType to PcdMeasuredBootHashMask

lsubash commented on PayloadPkg/OsLoader/OsLoader.c:
```diff
@@ -387,8 +387,7 @@ SetupBootImage (
         // Extend hash of the image into TPM.
-        TpmExtendPcrAndLogEvent (8, TPM_ALG_SHA256, LoadedImage->ImageHash,
-          EV_COMPACT_HASH, sizeof("LinuxLoaderPkg: OS Image"), (UINT8 *)"LinuxLoaderPkg: OS Image");
+        TpmExtendOsHash (LoadedImage->ImageHash, HASH_TYPE_SHA256);
```
Currently HASH_TYPE_SHA256 is used. Will improve to use the hash type from container format.

sagraw2 replied:
Instead of adding this new API here, I would recommend to file a separate patch and take care of complete change w.r.t OS hash.

lsubash replied:
TpmExtendPcrAndLogEvent() would enforce to use TPM Hash Alg. Also this API do not check for Active PCR banks. 
New API TpmExtendOsHash() would take care of additional checks.

mauricema commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -36,6 +36,61 @@ LoadComponentCallback (
+        return RETURN_NOT_FOUND;
+      }
+    } else {
+      Status = GetComponentInfo (ContainerSig, &Src, &Length);
```
why  do you need this ?    EPayload is always a container.    So if LocateComponentEntry() returned NULL, it is a failure condition already.

lsubash replied:
This case addresses if measured boot hash type differs from container hash type.
Eg: SM3 used for MB and RSA3K/SHA384 for signing process

sagraw2 replied:
Please add this explanation in the code comment appropriately so that future readers 
will know what/why is happening.

mauricema replied:
If so, we are recalc the hash using the info from GetComponentInfo.  This is not ideal.   The base will point to data in flash.  The process will be very slow.
During loading, it has been copied into memory already.  Might be better to do it at that time to avoid performance issue.  All TPM extension can occur at the same place where we do hash verification.  Any thoughts ?

sagraw2 replied:
Agreed. That is important to avoid reading from flash for hash calculation.

lsubash replied:
We are performing TPM operations for stage components(in stage1B and Stage2) immediately after LoadComponentWithCallback() in which hash verification happens. Yes, We could use the "memory Buffer addr" returned from LoadComponentWithCallback() in case HashCalculation need to be done.

mauricema replied:
Yes, maybe we can enhance the callback to support this.  Then the flow will be simpler to just extend hash in the callback function.

mauricema commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -36,6 +36,61 @@ LoadComponentCallback (
+  @retval RETURN_NOT_FOUND    EPayload hash Not found
+**/
+RETURN_STATUS
+ExtendEPayload (
```
ExtendEpayload

lsubash replied:
will do

mauricema commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -36,6 +36,61 @@ LoadComponentCallback (
+      if (ComponentEntry->HashSize == HashSize) {
+        TpmExtendHash (ComponentEntry->HashData, HashAlg);
+      } else {
+        DEBUG ((DEBUG_INFO, "EPAYLOAD hash does not exist !"));
```
EPAYLOAD hash is invalid !

lsubash commented:
With recent version fixed comments.

mauricema commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -36,6 +36,65 @@ LoadComponentCallback (
+  @retval RETURN_NOT_FOUND    EPayload hash Not found
+**/
+RETURN_STATUS
+ExtendEpayload (
```
Let us move it into Stage2Support.c file to make Stage2 main flow clean.

lsubash replied:
sure

gdong1 commented on BootloaderCommonPkg/Include/Library/ContainerLib.h:
```diff
@@ -33,7 +35,16 @@ typedef UINT8 AUTH_TYPE;
+  UINT32           CompLen;
+  HASH_ALG_TYPE    HashAlg;
+  UINT8           *HashData;
+} COMPONENT_CB_INFO;
```
Better use COMPONENT_CALLBACK_INFO if you have other things to update this patch.

mauricema commented on BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c:
```diff
@@ -778,7 +782,14 @@ LoadComponentWithCallback (
+    CbInfo.CompBuf          = CompBuf;
+    CbInfo.CompLen          = SignedDataLen;
+    CbInfo.HashAlg          = GetHashAlg(AuthType);
+    CbInfo.HashData         = HashData;
```
How do you know the HashData is valid since the Status returned by Auth() could be error ?  
Should we consider only extending hash when AuthenticateComponent() returns  SUCCESS ?

lsubash replied:
This callback would be invoked only for
- Stage2 authentication where hashdata would be NULL
- Payload and Epayload where hashdata is valid only for Epayload
Incase authentication fails, boot would halt. 
Added a additional check to invoke CB only Auth success cases

mauricema commented on BootloaderCorePkg/Stage2/Stage2.c:
```diff
@@ -34,6 +39,20 @@ LoadComponentCallback (
     break;
   }
+
+  if(CbInfo)
```
I think you should define a new function for this.  And only call this with  PROGESS_ID_AUTHENTICATE case.

lsubash replied:
done

mauricema commented on BootloaderCorePkg/Stage1B/Stage1B.c:
```diff
@@ -31,6 +34,16 @@ LoadComponentCallback (
   }
+
+  //Check the boot mode
+  BootMode = GetBootMode();
```
I think you should define a new function for this.  And only call this with  PROGESS_ID_AUTHENTICATE case.
